examples/step-50/step-50.cc

```diff
@@ -14,906 +14,1449 @@
-                                                         local_dof_indices[j],
-                                                         cell_matrix(i, j));
+              cell_matrix(i, j) +=
+                (coefficient_value * fe_values.shape_grad(i, q_point) *
+                 fe_values.shape_grad(j, q_point)) *
+                fe_values.JxW(q_point);
+
+        cell->get_mg_dof_indices(local_dof_indices);
+
+        boundary_constraints[cell->level()].distribute_local_to_global(
+          cell_matrix, local_dof_indices, mg_matrix[cell->level()]);
+
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          for (unsigned int j = 0; j < dofs_per_cell; ++j)
+            if (mg_constrained_dofs.is_interface_matrix_entry(
+                  cell->level(), local_dof_indices[i], local_dof_indices[j]))
+              mg_interface_in[cell->level()].add(local_dof_indices[i],
+                                                 local_dof_indices[j],
+                                                 cell_matrix(i, j));
+      }
+
+  for (unsigned int i = 0; i < triangulation.n_global_levels(); ++i)
+    {
+      mg_matrix[i].compress(VectorOperation::add);
+      mg_interface_in[i].compress(VectorOperation::add);
+    }
+}
+
+
+// @sect4{<code>LaplaceProblem::assemble_rhs()</code>}
+
+// This function assembles the right-hand side vector for the matrix-free
+// method. The function is similar to the one found in the ``Use
+// FEEvaluation::read_dof_values_plain() to avoid resolving constraints''
+// subsection in the ``Possibilities for extensions'' section of step-37.
+//
+// The reason for this function is that the MatrixFree operators do not take
+// into account non-homogeneous Dirichlet constraints, instead treating all
+// Dirichlet constraints as homogeneous. To account for this, the right-hand
+// side here is assembled as the residual $r_0 = f-Au_0$, where $u_0$ is a
+// zero vector except in the Dirichlet values. Then when solving, we have that
+// the solution is $u = u_0 + A^{-1}r_0$. This can be seen as a Newton
+// iteration on a linear system with initial guess $u_0$. The CG solve in the
+// solve() function below computes $A^{-1}r_0$ and the call to
+// `constraints.distribute()` (which directly follows) adds the $u_0$.
+//
+// Obviously, since we are considering a problem with zero Dirichlet boundary,
+// we could have taken a similar approach to step-37 assemble_rhs(), but this
+// additional work allows us to change the problem declaration if we so
+// choose.
+template <int dim, int degree>
+void LaplaceProblem<dim, degree>::assemble_rhs()
+{
+  TimerOutput::Scope timing(computing_timer, "Assemble right-hand side");
+
+  // The system_rhs vector is of type LA::MPI::Vector, but the MatrixFree
+  // class only work for dealii::LinearAlgebra::distributed::Vector.
+  // Therefore we must use the ChangeVectorType class above to copy between
+  // the two types.
```

Comment:
I guess it would make sense to add this block to the text above, to improve the text flow/code flow?

Reply:
I added two more paragraphs and removed the comments inside the function itself. Feel free to offer any suggestions on improved wording or unnecessary text etc.